presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -230,90 +230,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
-                            taskStatsCollector,
-                            toTaskProcessorBroadcastInputs(broadcastInputs)));
-        }
+        PrestoSparkTaskDescriptor taskDescriptor = new PrestoSparkTaskDescriptor(
+                session.toSessionRepresentation(),
+                session.getIdentity().getExtraCredentials(),
+                fragment,
+                tableWriteInfo);
+        SerializedPrestoSparkTaskDescriptor serializedTaskDescriptor = new SerializedPrestoSparkTaskDescriptor(
+                taskDescriptorJsonCodec.toJsonBytes(taskDescriptor));
 
+        Optional<Integer> numberOfInputPartitions = Optional.empty();
         ImmutableList.Builder<String> fragmentIds = ImmutableList.builder();
-        ImmutableList.Builder<RDD<Tuple2<Integer, PrestoSparkRow>>> rdds = ImmutableList.builder();
+        ImmutableList.Builder<RDD<Tuple2<Integer, PrestoSparkRow>>> inputRdds = ImmutableList.builder();
         for (Map.Entry<PlanFragmentId, JavaPairRDD<Integer, PrestoSparkRow>> input : rddInputs.entrySet()) {
             fragmentIds.add(input.getKey().toString());
-            rdds.add(input.getValue().rdd());
+            RDD<Tuple2<Integer, PrestoSparkRow>> rdd = input.getValue().rdd();
+            inputRdds.add(rdd);
+            if (!numberOfInputPartitions.isPresent()) {
+                numberOfInputPartitions = Optional.of(rdd.getNumPartitions());
+            }
+            else {
+                checkArgument(
+                        numberOfInputPartitions.get() == rdd.getNumPartitions(),
+                        "Incompatible number of input partitions: %s != %s",
+                        numberOfInputPartitions.get(),
+                        rdd.getNumPartitions());
+            }
         }
 
-        Function<List<Iterator<Tuple2<Integer, PrestoSparkRow>>>, Iterator<Tuple2<Integer, PrestoSparkRow>>> taskProcessor = createTaskProcessor(
+        PrestoSparkTaskProcessor taskProcessor = new PrestoSparkTaskProcessor(
                 executorFactoryProvider,
                 serializedTaskDescriptor,
                 fragmentIds.build(),
                 taskStatsCollector,
                 toTaskProcessorBroadcastInputs(broadcastInputs));
 
+        Optional<RDD<SerializedTaskSource>> taskSourceRdd;
+        List<TableScanNode> tableScans = findTableScanNodes(fragment.getRoot());
+        if (!tableScans.isEmpty()) {
+            PartitioningHandle partitioning = fragment.getPartitioning();
+            taskSourceRdd = Optional.of(createTaskSourcesRdd(sparkContext, session, partitioning, tableScans, numberOfInputPartitions).rdd());
+        }
+        else if (rddInputs.size() == 0) {
+            checkArgument(fragment.getPartitioning().equals(SINGLE_DISTRIBUTION), "SINGLE_DISTRIBUTION partitioning is expected: %s", fragment.getPartitioning());
+            taskSourceRdd = Optional.of(new PrestoSparkTaskSourceRdd(sparkContext.sc(), ImmutableList.of(ImmutableList.of())));
```

Comment:
We need to somehow create exactly 1 partition, so exactly one task is scheduled. Thats why we need this single partition input RDD. I agree that this is a bit of a hack, and in theory it is possible to create a single "dummy" partition of the `PrestoSparkTaskRdd` itself. But that might be a little messy, so I'm kinda trying to chose a lesser evil.

Reply:
@arhimondr : If the RDD becomes special enough in the case of `SINGLE_DISTRIBUTION` , a dedicated RDD might not be a bad idea, but you might have different opinions :) .

In any case, for now just explain in comment when the list is empty, it means exactly 1 partition (this is not intuitive to me), and perhaps a TODO about figure out less hacky approach? (We have 20% time assigned to Better Engineering).

Still , what about have a factory method such as `PrestoSparkTaskSourceRdd.createRddWithSingleDistribution` so you hide the detail that creating a `PrestoSparkTaskSourceRdd` with `ImmutableList.of(ImmutableList.of())` as task source?

Reply:
> @arhimondr : If the RDD becomes special enough in the case of SINGLE_DISTRIBUTION , a dedicated RDD might not be a bad idea, but you might have different opinions :) .

I would prefer to have a single way of creating a presto task. Otherwise we would have to maintain multiple ways. I think that's the main motivation to stick with  only `PrestoSparkTaskSourceRdd `, `PrestoSparkTaskRdd`.

> PrestoSparkTaskSourceRdd.createRddWithSingleDistribution

Then maybe `createRddWithSingleEmptyPartition`. But that might also be quite confusing. Let me leave a comment instead.